### PR TITLE
Add per site shield level settings

### PR DESCRIPTION
**Pull request by cuba**
Resolves https://github.com/brave/brave-browser/issues/39573

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1. Test all the existing functionality on the shields panel applies
2. Test shields panel UI across different platform and configuration, iPhone, iPad, dark mode, light mode, text sizes, iOS 16, 17 and 18 (if possible).
3. Test migrating from old toggle to the new selectors picker:
 - when disabled and migrating, the shield level is "Disabled"
 - when enabled before migrating, the shield level is "Standard"
4. Test that the global shield level is applied when in a site when the site shield level was not changed
5. Test P3A total shields lowered/increased from global

## Video
https://github.com/brave/brave-core/assets/909331/b7a130d9-c780-4997-90f3-5a23edb53e94

## Screenshots
Simulator Screenshot - iPhone 14 Pro - 2024-07-10 at 11 52 40
Simulator Screenshot - iPhone 14 Pro - 2024-07-10 at 11 52 42

### CODE REVIEW

File: ios/brave-ios/App/iOS/Delegates/AppDelegate.swift [hunk 1]
```diff
@@ -142,6 +142,9 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
   ) -> Bool {
     AppState.shared.state = .launching(options: launchOptions ?? [:], active: true)
 
+    // Run migrations that need access to Data
+    Migration.postDataLoadMigration()
```
Comment (Brandon-T) [inline ios/brave-ios/App/iOS/Delegates/AppDelegate.swift:146]: Why did you move this to be so early? This cannot run without `DataController.shared.initializeOnce()`

File: ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ToolbarDelegate.swift [hunk 2]
```diff
@@ -416,138 +416,150 @@ extension BrowserViewController: TopToolbarDelegate {
...
+      contentController: PopoverNavigationController(
+        rootViewController: ShieldsPanelViewController(
+          url: url,
+          tab: selectedTab,
+          domain: Domain.getOrCreate(forUrl: url, persistent: !selectedTab.isPrivate)
+        ) { [weak self, weak selectedTab] action in
+          switch action {
+          case .navigate(let target, let dismiss):
+            guard let self else { return }
+            guard let selectedTab else { return }
```
Comment (Brandon-T) [inline ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ToolbarDelegate.swift:440]: Combine the guard statements.

File: ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ToolbarDelegate.swift [hunk 3]
```diff
@@ -416,138 +416,150 @@ extension BrowserViewController: TopToolbarDelegate {
...
+    weak var spinner: SpinnerView?
+    let controller = UIHostingController(
+      rootView: AdvancedShieldsSettingsView(
+        settings: AdvancedShieldsSettings(
+          profile: self.profile,
+          tabManager: self.tabManager,
+          feedDataSource: self.feedDataSource,
+          historyAPI: self.braveCore.historyAPI,
+          p3aUtilities: self.braveCore.p3aUtils,
+          deAmpPrefs: self.braveCore.deAmpPrefs,
```
Comment (Brandon-T) [inline ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ToolbarDelegate.swift:519]: Pass in `braveCore` instead of all of these things, then access those classes inside your view.

File: ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ToolbarDelegate.swift [hunk 4]
```diff
@@ -416,138 +416,150 @@ extension BrowserViewController: TopToolbarDelegate {
...
+          debounceService: DebounceServiceFactory.get(privateMode: false),
```
Comment (Brandon-T) [inline ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ToolbarDelegate.swift:520]: `DebounceServiceFactory` doesn't require a `braveCore` instance?

File: ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+      }
+      .padding(.top)
+    }
+    .background(Color(.braveBackground))
+    .frame(idealWidth: 360, alignment: .center)
+    .task {
+      if let faviconImage = FaviconFetcher.getIconFromCache(for: url)?.image {
+        self.favicon = Image(uiImage: faviconImage)
+      }
+      await loadFavicon(for: url, isPrivateBrowsing: viewModel.isPrivateBrowsing)
```
Comment (Brandon-T) [inline ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift:80]: There is a favicon SwiftUI view already. Use that. Wallet and AI-Chat both use it.

File: ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+          .frame(width: 24, height: 24, alignment: .center)
+        Text(displayHost)
+          .font(.title2)
+          .foregroundStyle(Color(.bravePrimary))
+      }
+      .frame(minWidth: .zero, alignment: .center)
+
+      ShieldsSwitchView(isEnabled: $viewModel.shieldsEnabled)
+        .frame(
+          width: ShieldsSwitch.size.width,
```
Comment (Brandon-T) [inline ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift:101]: Why do we hard-code the size of the view here?

File: ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+
+  var body: some View {
+    VStack(alignment: .leading, spacing: 8) {
+      Text(title)
+        .font(.footnote)
+        .foregroundStyle(Color(.secondaryBraveLabel))
+        .textCase(.uppercase)
+        .padding(.horizontal)
+        .padding(.top, 8)
+        .padding(.bottom, 0)
```
Comment (Brandon-T) [inline ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift:346]: Why do you need padding of zero? Is this line necessary?

File: ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+            }
+            .foregroundStyle(Color(.secondaryBraveLabel))
+            .font(.body)
+          }
+          .frame(maxWidth: .infinity, alignment: .center)
+          .padding()
+          .contentShape(Rectangle())
+        }
+      }
+      .padding(0)
```
Comment (Brandon-T) [inline ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift:403]: Why do you need `.padding(0)`?

File: ios/brave-ios/Tests/DataTests/DomainTests.swift [hunk 9]
```diff
@@ -241,3 +238,22 @@ class DomainTests: CoreDataTestCase {
     XCTAssertFalse(raydiumDomain.walletPermissions(for: .sol, account: walletSolAccount))
   }
 }
+
+extension Domain {
+  fileprivate class func performChangesOnDomain(
```
Comment (Brandon-T) [inline ios/brave-ios/Tests/DataTests/DomainTests.swift:243]: Why is this function necessary? It's only used in Unit Tests.

**Review by Brandon-T — CHANGES_REQUESTED** [review]

File: ios/brave-ios/App/iOS/Delegates/AppDelegate.swift (hunk 1 above)
Comment (cuba) [reply ios/brave-ios/App/iOS/Delegates/AppDelegate.swift:146]: Honestly I wasn't sure where to put this, it certainly can move. I can put it in the setter like the other migrations, but setting `AppState.shared.state = .launching(options: launchOptions ?? [:], active: true)` calls `DataController.shared.initializeOnce()`

File: ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ToolbarDelegate.swift (hunk 4 above)
Comment (cuba) [reply ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ToolbarDelegate.swift:520]: Nope, it's a singleton factory class as other examples that use these factories

File: ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift (hunk 5 above)
Comment (cuba) [reply ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift:80]: Perfect, thanks!

File: ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift (hunk 6 above)
Comment (cuba) [reply ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift:101]: just because its already hard-coded in that existing class so I wanted it to match. I don't remember if there were any specific reasons other than that but I can see what happens when I remove this

File: ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift (hunk 7 above)
Comment (cuba) [reply ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift:346]: Just so there is no space between the advanced header and bottom when advanced section is collapsed as was the original design.

File: ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift (hunk 8 above)
Comment (cuba) [reply ios/brave-ios/Sources/Brave/Frontend/Shields/ShieldsPanelView.swift:403]: so i can stretch certain elements, such as dividers and the headers left and right fully as is in the original screen.

File: ios/brave-ios/Tests/DataTests/DomainTests.swift (hunk 9 above)
Comment (cuba) [reply ios/brave-ios/Tests/DataTests/DomainTests.swift:243]: This is only for tests we need to ensure changes are saved before running tests. There was already somethiing like this but it was specific to changing one of the BraveShield enums. 

Since I removed `adblockAndTp` from that enum, I needed to create another method to do this so I created a more general one and privatized it to this file. 

Actually that existing function was only used in this test, so I can remove it and put it here:

```
extension Domain {
  private func setBraveShield(
    shield: BraveShield,
    isOn: Bool?,
    context: NSManagedObjectContext
  ) {

    let setting = (isOn == shield.globalPreference ? nil : isOn) as NSNumber?
    switch shield {
    case .allOff: shield_allOff = setting
    case .fpProtection: shield_fpProtection = setting
    case .noScript: shield_noScript = setting
    }
  }
  
  fileprivate class func setBraveShield(
    forUrl url: URL,
    shield: BraveShield,
    isOn: Bool?,
    isPrivateBrowsing: Bool
  ) {
    let _context: WriteContext = isPrivateBrowsing ? .new(inMemory: true) : .new(inMemory: false)
    setBraveShieldInternal(forUrl: url, shield: shield, isOn: isOn, context: _context)
  }
  
  fileprivate class func setBraveShieldInternal(
    forUrl url: URL,
    shield: BraveShield,
    isOn: Bool?,
    context: WriteContext = .new(inMemory: false)
  ) {
    DataController.perform(context: context) { context in
      // Not saving here, save happens in `perform` method.
      let domain = Domain.getOrCreateInternal(
        url,
        context: context,
        saveStrategy: .delayedPersistentStore
      )
      domain.setBraveShield(shield: shield, isOn: isOn, context: context)
    }
  }
  
  fileprivate class func performChangesOnDomain(
    for url: URL,
    isPrivateBrowsing: Bool,
    changes: @escaping (Domain) -> Void
  ) {
    let context = WriteContext.new(inMemory: isPrivateBrowsing)
    DataController.perform(context: context) { context in
      // Not saving here, save happens in `perform` method.
      let domain = Domain.getOrCreateInternal(
        url,
        context: context,
        saveStrategy: .delayedPersistentStore
      )
      changes(domain)
    }
  }
}
```

And of course, I can slim this down and remove some of the duplicated unecessary stuff, but all the functions in the above (except for the last one) are existing functions which were only used in these tests and I wanted to show this.

Comment (cuba) [reply ios/brave-ios/Tests/DataTests/DomainTests.swift:243]: If you would like to see a commit i can do that. but in the end these are the changes:

Screenshot 2024-07-09 at 4 48 53 PM
Screenshot 2024-07-09 at 4 49 05 PM

File: ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ToolbarDelegate.swift (hunk 3 above)
Comment (cuba) [reply ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ToolbarDelegate.swift:519]: Ok but FYI, This will result in more file changes as the signature of this initializer is not in the scope of these changes.

**Review by Brandon-T — APPROVED** [review]
